internal/statemachine/helper.go

```diff
@@ -894,3 +894,108 @@ func checkCustomizationSteps(searchStruct interface{}, tag string) (extraStates
 	}
 	return extraStates
 }
+
+// updateGrub mounts the resulting image and runs update-grub
+func (stateMachine *StateMachine) updateGrub(rootfsVolName string, rootfsPartNum int) error {
+	// find a /dev/loop* that is not already in use
+	freeDevLoop := 0
+	foundFree := false
+	for !foundFree {
+		freeDevLoop++
+		loops, err := filepathGlob(fmt.Sprintf("/dev/loop%d*", freeDevLoop))
+		if err != nil {
+			return fmt.Errorf("Error globbing for /dev/loop%d: \"%s\"", freeDevLoop, err.Error())
+		}
+		if len(loops) == 0 {
+			foundFree = true
+		}
+	}
+
+	// create a directory in which to mount the rootfs
+	mountDir := filepath.Join(stateMachine.tempDirs.scratch, "loopback")
+	err := osMkdir(mountDir, 0755)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("Error creating scratch/loopback directory: %s", err.Error())
+	}
+
+	// Slice used to store all the commands that need to be run
+	// to properly update grub.cfg in the chroot
+	var updateGrubCmds []*exec.Cmd
+
+	imgPath := filepath.Join(stateMachine.commonFlags.OutputDir, stateMachine.VolumeNames[rootfsVolName])
+
+	// set up the loopback
+	var umounts []*exec.Cmd
+	updateGrubCmds = append(updateGrubCmds,
+		[]*exec.Cmd{
+			// set up the loopback
+			exec.Command("losetup",
+				filepath.Join("/dev", fmt.Sprintf("loop%d", freeDevLoop)),
+				imgPath,
+			),
+			exec.Command("kpartx",
+				"-a",
+				filepath.Join("/dev", fmt.Sprintf("loop%d", freeDevLoop)),
+			),
+			// mount the rootfs partition in which to run update-grub
+			exec.Command("mount",
+				filepath.Join("/dev",
+					"mapper",
+					fmt.Sprintf("loop%dp%d", freeDevLoop, rootfsPartNum),
+				),
+				mountDir,
+			),
+		}...,
+	)
+
+	// set up the mountpoints
+	mountPoints := []string{"/dev", "/proc", "/sys"}
+	for _, mountPoint := range mountPoints {
+		mountCmd, umountCmd := mountFromHost(mountDir, mountPoint)
+		updateGrubCmds = append(updateGrubCmds, mountCmd)
+		umounts = append(umounts, umountCmd)
+		defer umountCmd.Run()
+	}
+	// make sure to unmount the disk too
+	umounts = append(umounts, exec.Command("umount", mountDir))
+
+	// actually run update-grub
+	updateGrubCmds = append(updateGrubCmds,
+		exec.Command("chroot",
+			mountDir,
+			"update-grub",
+		),
+	)
+
+	// unmount /dev /proc and /sys
+	updateGrubCmds = append(updateGrubCmds, umounts...)
```

Comment:
Haven't the umounts already been deferred?

Reply:
Yes they have, but calling them this way gives us error checking.

Reply:
OK